frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,372 @@
+
+        end
+    end
+    return val
+end
+
+-------------------------------------
+-- dictionary, file functions use:
+--    self.dict_file
+--    self.new_dict_file
+--    self.dict
+--    self.next_line
+-------------------------------------
+
+-- opens user dictionary
+function ReaderUserHyph:openDictionary()
+    self.dict_file = self.ui.userhyph:getDictionaryPath()
+    self.new_dict_file = self.dict_file .. ".new"
+
+    -- if no file, crate an empty one
+    if lfs.attributes(self.dict_file, "mode") ~= "file" then
+        self.dict = io.open(self.dict_file, "w")
+        self.dict:close()
+    end
+
+    -- check user dict file
+    if string.sub(lfs.attributes(self.dict_file).permissions, 1, 2) ~= "rw" then
+        UIManager:show(InfoMessage:new{
+            text = T(_("Wrong file permissions for:\n\"%1\""), self.dict_file),
+            show_icon = true,
+        })
+        return
+    end
+
+    -- open files
+    self.dict = io.open(self.dict_file, "r")
+    if not self.dict then
+        logger.err("UserHyph: could not open " .. self.dict_file)
+        return
+    end
+    self.new_dict = io.open(self.new_dict_file, "w")
+    if not self.dict then
+        logger.err("UserHyph: could not open " .. self.new_dict_file)
+        return
+    end
+    return true
+end
+
+-- reads the first dictionary entries from the old dictionary
+-- and write them to the new dict, as long the entry is alphabetically lower than word
+-- if the word matches the dictionary entry, the hypenation part is formatted and returned
+function ReaderUserHyph:findEntry(word)
+    -- scan hyphenation dictionary for selected word
+    local word_lower = Utf8:lower(word)
+    self.next_line = self.dict:read()
+    while self.next_line and Utf8:lower(self.next_line:sub(1, self.next_line:find(";") - 1)) < word_lower do
+        self.new_dict:write(self.next_line .. "\n")
+        self.next_line = self.dict:read()
+    end
+
+    local hyphenation
+    -- check if a hyphenation is found for word
+    if self.next_line and Utf8:lower(self.next_line:sub(1, self.next_line:find(";") - 1)) == word_lower then
+        hyphenation = string.sub(self.next_line, self.next_line:find(";") + 1) -- hyphenation found
+        self.next_line = nil -- Important for not duplicating the entry in the dictionary
+    end
+    return hyphenation
+end
+
+-- writes on entry to new dictionary
+function ReaderUserHyph:writeEntry(line)
+    self.new_dict:write(Utf8:lower(line) .. "\n")
+end
+
+-- reads the rest of the old dictionary and writes this to the reset
+-- if self.new_dict~=nil, an additional line can be written
+function ReaderUserHyph:writeRest()
+    -- write old entry if there was one
+    if self.next_line then
+        self.new_dict:write(self.next_line .. "\n")
+    end
+    -- copy rest of file
+    repeat
+        self.next_line = self.ui.userhyph.dict:read()
+        if self.next_line then
+            self.new_dict:write(self.ui.userhyph.next_line .. "\n")
```

Comment:
I understand all this, but it feels a bit odd having file updating split into multiple little functions, having file handles as `self.new_dict`.
Also, feels overkill (and too low level for my liking) doing it that way :) I would have been lazy and load the old file into a Lua table key/value, and just dump this table to file after sorting by key.

Reply:
That would mean to have the dictionary once more in the memory (beside of crengine).
I would stay with that, it is well documented.

Reply:
You forced me to really have a look at it :)
Yes, it's documented, but it feels a bit not natural or over-engineered (your 2nd middle name :)
When the user is to add/edit a word:
- you read the file up to the position this word should be, getting the previous value to allow the user to edit it - while keeping the file opened so we can read the "rest" from there
- doing that, you create the file.new, copying all the content up to there (which feels like an optimisation, or a waste if the user cancels the edit)
- then, in callbacks, you continue the work on these 2 file handles, writing the new entry, copying the rest - or cancel it, removing the new file

I don't think we have anything doing that kind of work in the current code base - so it's something new, but this method feels a bit odd.
Dunno how @NiLuJe and @Frenzie you feel about this. (I don't mind much the way it works as this feature will be rarely used by only a few users, but I surely don't want to have to maintain it :)
One idea would be to at least not create the new file until really needed, use tell() and seek() to remember the position at which we found the word to have the info to quickly do the copy + addition + copy rest when really storing the value. Other ideas ?
(But it still feels over engineered for this so rarely used feature that a plain findEntry() reading and closing - and a updateEntry() reading again the whole file, and rewritting it from scratch would probably takes a lot less lines of code.)

Reply:
Haven't read through it, but what's the format? It looks like a midly formatted plaintext with comma separators or some such?

I'd quite simply *not* deal with a string text storage, and just eat the cost of converting that to a native data format (i.e., a hash map?).

Reply:
It's plain text, one line per item:
```
nombre;no-m-bre
souabes;sou-abes
```
and it is to be parsed by frontend (for storing/updating individual items via UI) and by crengine (so it also get this key>value structure in a similar way it already is used for the other hyphenation that it reads from XML files).
And I guess we want it readable/editable by hand, so a hyphenation lover could edit it or generate it from other sources.

Reply:
I guess it's fine? I wouldn't personally have gone quite this in-depth on it unless I had a 100k line file and was noticing some performance issue, but eh. :-P